tensornetwork/tn_keras/mpo.py

```diff
@@ -0,0 +1,168 @@
+    super(DenseMPO, self).__init__(**kwargs)
+
+  def build(self, input_shape: List[int]) -> None:
+    # Disable the attribute-defined-outside-init violations in this function
+    # pylint: disable=attribute-defined-outside-init
+    if input_shape[-1] is None:
+      raise ValueError('The last dimension of the inputs to `Dense` '
+                       'should be defined. Found `None`.')
+
+    def is_perfect_root(n, n_nodes):
+      root = n**(1. / n_nodes)
+      return round(root)**n_nodes == n
+
+    # Ensure the MPO dimensions will work
+    assert is_perfect_root(input_shape[-1], self.num_nodes)
+    assert is_perfect_root(self.output_dim, self.num_nodes)
+
+    self.in_leg_dim = int(input_shape[-1]**(1. / self.num_nodes))
+    self.out_leg_dim = int(self.output_dim**(1. / self.num_nodes))
+
+    self.nodes.append(
+        self.add_weight(name='end_node_first',
+                        shape=(self.in_leg_dim, self.bond_dim,
+                               self.out_leg_dim),
+                        trainable=True,
+                        initializer=self.kernel_initializer))
+    for i in range(self.num_nodes - 2):
+      self.nodes.append(
+          self.add_weight(name=f'middle_node_{i}',
+                          shape=(self.in_leg_dim, self.bond_dim, self.bond_dim,
+                                 self.out_leg_dim),
+                          trainable=True,
+                          initializer=self.kernel_initializer))
+    self.nodes.append(
+        self.add_weight(name='end_node_last',
+                        shape=(self.in_leg_dim, self.bond_dim,
+                               self.out_leg_dim),
+                        trainable=True,
+                        initializer=self.kernel_initializer))
+
+    self.bias_var = self.add_weight(
+        name='bias',
+        shape=(self.output_dim,),
+        trainable=True,
+        initializer=self.bias_initializer) if self.use_bias else None
+    super(DenseMPO, self).build(input_shape)
+
+  def call(self, inputs: tf.Tensor) -> tf.Tensor:
+
+    def f(x: tf.Tensor, nodes: List[Node], num_nodes: int, in_leg_dim: int,
+          use_bias: bool, bias_var: tf.Tensor) -> tf.Tensor:
+      orig_shape = x.shape
+      l = [in_leg_dim] * num_nodes
+      input_reshaped = tf.reshape(x, tuple(l))
+      x_node = tn.Node(input_reshaped, name='xnode', backend="tensorflow")
+
+      tn_nodes = []
+      for i, v in enumerate(nodes):
+        tn_nodes.append(tn.Node(v, name=f'node_{i}', backend="tensorflow"))
+        # Connect every node to input node
+        x_node[i] ^ tn_nodes[i][0]
+
+      for i, _ in enumerate(tn_nodes):
+        try:
+          tn_nodes[i][2] ^ tn_nodes[i + 1][2]
+        except: # pylint: disable=W0702
+          pass
+        try:
+          tn_nodes[i][1] ^ tn_nodes[i + 1][1]
+        except: # pylint: disable=W0702
+          pass
+        try:
+          tn_nodes[i][2] ^ tn_nodes[i + 1][1]
+        except: # pylint: disable=W0702
+          pass
```

Comment:
The naked try except is a little sketchy, especially since you're disabling the linter on this. If there is a real error connecting the legs (perhaps the dimensions were mismatched?) this would eat the error.

I assume you do this to account the the end nodes. Can you refactor this to not use try/excepts?